src/Database/Esqueleto/Record.hs

```diff
@@ -0,0 +1,423 @@
+  -- Pairs of record field names and local variable names.
+  fieldNames <- forM sqlFields (\(name', _type) -> do
+    var <- newName $ nameBase name'
+    pure (name', var))
+
+  -- Patterns binding record fields to local variables.
+  let fieldPatterns :: [FieldPat]
+      fieldPatterns = [(name', VarP var) | (name', var) <- fieldNames]
+
+      -- Local variables for fields joined with `:&` in a single expression.
+      joinedFields :: Exp
+      joinedFields =
+        case snd `map` fieldNames of
+          [] -> TupE []
+          [f1] -> VarE f1
+          f1 : rest ->
+            let helper lhs field =
+                  InfixE
+                    (Just lhs)
+                    (ConE '(:&))
+                    (Just $ VarE field)
+             in foldl' helper (VarE f1) rest
+
+  identInfo <- newName "identInfo"
+  -- Roughly:
+  -- sqlSelectCols $identInfo SqlFoo{..} = sqlSelectCols $identInfo $joinedFields
+  pure $
+    FunD
+      'sqlSelectCols
+      [ Clause
+          [ VarP identInfo
+          , RecP sqlName fieldPatterns
+          ]
+          ( NormalB $
+              AppE
+                (AppE (VarE 'sqlSelectCols) (VarE identInfo))
+                (ParensE joinedFields)
+          )
+          -- `where` clause.
+          []
+      ]
+
+-- | Generates the `sqlSelectColCount` declaration for an `SqlSelect` instance.
+sqlSelectColCountDec :: RecordInfo -> Q Dec
+sqlSelectColCountDec RecordInfo {..} = do
+  let joinedTypes =
+        case snd `map` sqlFields of
+          [] -> TupleT 0
+          t1 : rest ->
+            let helper lhs ty =
+                  InfixT lhs ''(:&) ty
+             in foldl' helper t1 rest
+
+  -- Roughly:
+  -- sqlSelectColCount _ = sqlSelectColCount (Proxy @($joinedTypes))
+  pure $
+    FunD
+      'sqlSelectColCount
+      [ Clause
+          [WildP]
+          ( NormalB $
+              AppE (VarE 'sqlSelectColCount) $
+                ParensE $
+                  AppTypeE
+                    (ConE 'Proxy)
+                    joinedTypes
+          )
+          -- `where` clause.
+          []
+      ]
+
+-- | Generates the `sqlSelectProcessRow` declaration for an `SqlSelect`
+-- instance.
+sqlSelectProcessRowDec :: RecordInfo -> Q Dec
+sqlSelectProcessRowDec RecordInfo {..} = do
+  -- Binding statements and field expressions (used in record construction) to
+  -- fill out the body of the main generated `do` expression.
+  --
+  -- Each statement is like:
+  --     Value fooName' <- takeColumns @(SqlExpr (Value Text))
+  -- A corresponding field expression would be:
+  --     fooName = fooName'
+  --
+  -- See `sqlSelectProcessRowPat` for the left-hand side of the patterns.
+  (statements, fieldExps) <-
+    unzip <$> forM (zip fields sqlFields) (\((fieldName', fieldType), (_, sqlType')) -> do
+      valueName <- newName (nameBase fieldName')
+      pure
+        ( BindS
+            (sqlSelectProcessRowPat fieldType valueName)
+            (AppTypeE (VarE 'takeColumns) sqlType')
```

Comment:
So, you *would* need to export `takeColumns` (and then have it be imported at the use-sites of the esqueleto), *if* you wrote `mkName "takeColumns"` instead of `'takeColumns`. But this `'valueName` style gives you a globally qualified name that does not need to be imported in the file that uses the macro.

Reply:
`mkName` gives you a dynamically bound variable, while `'valueName` gives you a name referring to the term `valueName` in scope at the *definition* of the macro.

Yesod actually does this when making the handlers - it does `mkName (verb ++ handlerName)`, and then *expects* you to import/define the relevant function so it can use that in the splice site.